beeline.go

```diff
@@ -97,65 +102,146 @@ func Init(config Config) {
-// context should also be responsible for sending that event on to Honeycomb
-// when it's finished.
-func ContextWithEvent(ctx context.Context, ev *libhoney.Event) context.Context {
-	return context.WithValue(ctx, honeyEventContextKey, ev)
+// AddFieldToSpan allows you to add a single field to an event anywhere
+// downstream of an instrumented request. After adding the appropriate
+// middleware or wrapping a Handler, feel free to call AddFieldToSpan freely
+// within your code. Pass it the context from the request (`r.Context()`) and
+// the key and value you wish to add.This function is good for span-level data,
+// eg timers or the arguments to a specific function call, etc. Fields added
+// here are prefixed with `app.`
+func AddFieldToSpan(ctx context.Context, key string, val interface{}) {
+	namespacedKey := fmt.Sprintf("app.%s", key)
+	span := internal.CurrentSpan(ctx)
+	if span != nil {
+		span.AddField(namespacedKey, val)
+	}
 }
 
-// ContextEvent retrieves the Honeycomb event from a context. You can add fields
-// to the event or override settings (eg sample rate) but should not Send() the
-// event; the wrapper that inserted the event into the Context is responsible
-// for sending it to Hnoeycomb
-func ContextEvent(ctx context.Context) *libhoney.Event {
-	if ctx != nil {
-		if evt, ok := ctx.Value(honeyEventContextKey).(*libhoney.Event); ok {
-			return evt
-		}
+// AddRollupFieldToSpan allows you to add a numeric field to the current span,
+// and, when called on multiple spans within a trace, the sum of the field will
+// be added to the root span. Use this when doing an action many times or on
+// many spans and you want the sum of all those actions to be represented on the
+// root span. Fields added here are prefixed with `app.` and the rolled up
+// fields on the root span are prefixed with `totals.app.`
+func AddRollupFieldToSpan(ctx context.Context, key string, val float64) {
+	namespacedKey := fmt.Sprintf("app.%s", key)
+	span := internal.CurrentSpan(ctx)
+	if span != nil {
+		span.AddRollupField(namespacedKey, val)
 	}
-	return nil
 }
 
-// contextBuilder isn't used yet but matches ContextEvent. When it's useful,
-// export it, but until then it's just confusing.
-func contextBuilder(ctx context.Context) *libhoney.Builder {
-	if bldr, ok := ctx.Value(honeyBuilderContextKey).(*libhoney.Builder); ok {
-		return bldr
+// AddFieldToTrace adds the field to both the currently active span and all
+// other spans involved in this trace that occur within this process.
+// Additionally, these fields are packaged up and passed along to downstream
+// processes if they are also using a beeline. This function is good for adding
+// context that is better scoped to the request than this specific unit of work,
+// eg user IDs, globally relevant feature flags, errors, etc. Fields added here
+// are prefixed with `global.`
+func AddFieldToTrace(ctx context.Context, key string, val interface{}) {
```

Comment:
I find the terminology here a bit awkward. As an application developer, the term "global" connotes _process-global_ (as in "global variable"), so I'd think of "global fields" as meaning "shared across _all requests_" (e.g. hostname, app configs). Whereas here we're defining it to mean "shared across all the spans in a _single request_".

I'd expect a good proportion of people to be confused by this (and, e.g., not realise they can add something like a team ID here, because it doesn't make sense for a team ID to be process-global. I'm also not sure the querying experience is very intuitive - `P95(durationMs) BREAK DOWN BY global.userID` would trip me up thinking "what does it mean to have a _global_ user ID?".

I'd prefer a prefix like `trace.` except of course that's reserved for trace meta-fields.

Ideas:
 * just use `app.`?
 * `request.`? Would only make sense for web apps, of course, and we might choose to worry about the potential for clobbering the fields from the request instrumentation.
 * what if whoever started the trace (by definition someone using the more-sophisticated API) picked a prefix for trace-level fields? e.g. an HTTP server middleware could set the prefix to be `request.`, a cron job wrapper or job queue consumer plugin could set it to be `job.`, etc.

Reply:
the node beeline uses `app.` and doesn't differentiate (at the column name level) between span and trace level fields. I do appreciate that makes for one place to find things, but I also appreciate the signal that I can use this same field in filtering other types of spans.  For consistency, I'll go with `app.` to match the node beeline.